### PR TITLE
Case search `proximity()` function

### PR DESCRIPTION
**Pull request by esoergel**
## Product Description
https://dimagi-dev.atlassian.net/browse/USH-1791
Pairs up with https://github.com/dimagi/commcare-hq/pull/31402

This adds a new function to case search to allow filtering gps case properties by proximity.  I'm soliciting input on a new name in the ticket.  Usage looks like

> proximity("coordinates", "42.4402967 -71.1453275", 1, "miles")

This expression will match all cases with a "coordinates" case property (defined as a GPS property in the DD) within 1 mile of 42.4402967, -71.1453275.

The arguments are

* The GPS case property
* The reference location
* Distance to search
* Distance units (one of "miles, yards, feet, inch, kilometers, meters, centimeters, millimeters, nauticalmiles")

## Technical Summary

by-commit should make sense

## Feature Flag
Though not specifically gated, it's useless without the `CASE_SEARCH_SMART_TYPES` flag:
> USH: Intelligently index specific case properties using the data dictionary

### Safety story

Lotsa automated tests

### Automated test coverage

Included

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by orangejenny** [comment]
haven't read this yet, but pleased that it supports nautical miles

File: corehq/apps/case_search/xpath_functions/utils.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+from corehq.apps.case_search.exceptions import XPathFunctionException
+
+
+def confirm_args_count(node, expected):
+    actual = len(node.args)
+    if actual != expected:
+        raise XPathFunctionException(
+            _(f'The "{node.name}" function accepts exactly {expected} arguments, got {actual}'),
+            serialize(node)
+        )
```
Comment (esoergel) [inline corehq/apps/case_search/xpath_functions/utils.py:14]: I initially had aims of putting more such validator functions in here, like confirming that arguments are of the appropriate type (numbers, strings), but didn't want to drop the function-specific context ("the second argument must be a number")

Comment (snopoke) [reply corehq/apps/case_search/xpath_functions/utils.py:14]: I've also wanted to do this. I've been toying with the idea of a decorator that specifies the argument names and types which can be used for validation and type casting.

**Review by snopoke — APPROVED** [review]

**Review by proteusvacuum — APPROVED** [review]

File: corehq/apps/case_search/xpath_functions/utils.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/case_search/xpath_functions/utils.py:14]: That'd be cool - I thought about doing something like argparse, where you implement a class for each command, but I'm loathe to introduce another inheritance structure.  A decorator would be a nice way to go about it.

**Review by joxl — COMMENTED** [review]
Elastic-related portion LGTM. ~I'll defer full review to @proteusvacuum~ (oops, you already got that :+1: )

**Comment by esoergel** [comment]
Gonna merge this as-is so we can start testing it out, and change the name separately if we decide to.

**Pull request merged**